pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1004,6 +1052,57 @@ BPF_PERF_OUTPUT(net_events, 1024);  // network events submission
 
 static __always_inline void *get_path_str(struct path *path);
 
+// HELPERS: ERRORS ---------------------------------------------------------------------------------
+
+static __always_inline void do_tracee_error(
+    void *ctx, enum bpf_error_level level, enum bpf_error_id id, s64 ret, u32 line, void *file)
+{
+    if (!ctx || !file)
+        return;
+
+    u32 zero = 0;
+    bpf_error_output_t *err_output = bpf_map_lookup_elem(&error_output_scratch, &zero);
+    if (unlikely(err_output == NULL))
+        return;
+
+    err_output->level = level;
+    err_output->id = id;
+
+    err_output->err.ret = ret;
+    err_output->err.cpu = bpf_get_smp_processor_id();
+    err_output->err.line = line;
+    u64 fsize = __builtin_strlen(file);
+    if (unlikely(fsize >= BPF_MAX_ERR_FILE_LEN))
+        fsize = BPF_MAX_ERR_FILE_LEN - 1;
+    __builtin_memcpy(err_output->err.file, file, fsize);
+    err_output->err.file[fsize] = '\0';
+
+    bpf_error_count_t counter_buf = {};
+    counter_buf.count = 1;
+    counter_buf.ts = bpf_ktime_get_ns(); // store the current ts
+    u64 ts_prev = 0;
+
+    bpf_error_count_t *counter = bpf_map_lookup_elem(&errors_count, &err_output->err);
+    if (likely(counter != NULL)) {
+        ts_prev = counter->ts; // store previous ts
+
+        counter->count += 1;
+        counter->ts = counter_buf.ts; // set to current ts
+    } else {
+        counter = &counter_buf;
+        bpf_map_update_elem(&errors_count, &err_output->err, counter, BPF_ANY);
+    }
+
+    // submit error when its cpu occurrence time diff is greater than 2s
+    if ((counter->ts - ts_prev) > (u64) 2000000000) {
+        err_output->count = counter->count;
```

Comment:
In this case, being a per-CPU map makes this atomic (as no other CPU will be changing the value pointed by the err_output pointer.